mp4e/mp4.go

```diff
@@ -0,0 +1,308 @@
+	SampleDurationMax uint64        // longest sample duration
+	SampleDurationMin uint64        // shortest sample duration
+	Segments          []SegmentInfo //
+	Size              uint64        // file size
+	Timescale         uint32        //
+}
+
+func (s *Mp4Info) AddError(err string, seg, frag int, pts int64) {
+	seg++
+	frag++
+
+	if seg > 0 {
+		err += fmt.Sprintf(", segment %d", seg)
+	}
+	if frag > 0 {
+		err += fmt.Sprintf(", fragment %d", frag)
+	}
+	if pts >= 0 {
+		err += fmt.Sprintf(", pts %d", pts)
+	}
+
+	s.Errors = append(s.Errors, err)
+}
+
+func (s *Mp4Info) String() string {
+	var sb strings.Builder
+	if len(s.Errors) > 0 {
+		b, _ := json.MarshalIndent(s.Errors, "", "  ")
+		_, _ = fmt.Fprintf(&sb, "errors: %s\n", string(b))
+	}
+	_, _ = fmt.Fprintf(&sb,
+		"size: %d\ntimescale: %d\nfragments: %d, samples/frames: %d\nsample duration range: %d - %d\nsamples per segment: %d - %d\n\n",
+		s.Size, s.Timescale,
+		s.FragmentCount, s.SampleCount,
+		s.SampleDurationMin, s.SampleDurationMax,
+		s.SampleCountMin, s.SampleCountMax,
+	)
+	for i, seg := range s.Segments {
+		_, _ = fmt.Fprintf(&sb, "%d: dts: %d - %d, pts %d - %d, seq %d - %d, samples: %d\n",
+			i+1, seg.DtsStart, seg.DtsEnd, seg.PtsStart, seg.PtsEnd, seg.SeqStart, seg.SeqEnd, seg.SampleCount)
+	}
+	return sb.String()
+}
+
+type InitInfo struct {
+}
+type SegmentInfo struct {
+	DtsEnd        uint64
+	DtsStart      uint64
+	FragmentCount uint64
+	PtsEnd        uint64
+	PtsStart      uint64
+	SampleCount   uint64
+	Samples       []*SampleInfo
+	SeqEnd        uint32
+	SeqStart      uint32
+}
+
+type SampleInfo struct {
+	Duration uint64
+	Fragment int
+	Pts      uint64
+}
+
+func ValidateFmp4(reader io.Reader) (file *mp4.File, info *Mp4Info, err error) {
+	e := errors.Template("mp4e.ValidateFmp4", errors.K.Invalid.Default())
+	if file, err = mp4.DecodeFile(reader); err != nil {
+		err = e(err)
+		return
+	}
+
+	info = &Mp4Info{
+		Size: file.Size(),
+	}
+
+	if file.Init != nil {
+		// moov - container for all the metadata
+		//   trak - container for an individual track or stream
+		//     tkhd - track header, overall information about the track; height, width
+		//     mdia - container for the media information in a track
+		//       mdhd - media header, overall information about the media; timescale
+		//       minf - media information container
+		//         stbl - sample table box, container for the time/space map
+		//           stsd - sample descriptions (codec types, initialization etc.); codec info
+		// file.Moov.Trak.Mdia.Minf.Stbl.Stsd.AvcX.AvcC
+		info.InitInfo = &InitInfo{}
+	}
+	if len(file.Segments) == 0 {
+		if file.Init == nil {
+			info.AddError("No segments", -1, -1, -1)
+		}
+		return
+	}
+
+	if !file.IsFragmented() {
+		info.AddError("Not fragmented", -1, -1, -1)
+	}
+	if file.Segments[0].Sidx != nil {
+		info.Timescale = file.Segments[0].Sidx.Timescale
+	}
+
+	dtsPrev := uint64(0)
+	ptsPrev := uint64(0)
```

Comment:
Why is this a uint64, but PTS is (I believe) int64?

Reply:
Mainly because the mp4ff library uses uint64. I tend to not like using unsigned numbers because it can hide errors, e.g. if some math produces a negative output when you're not expecting it then it's pretty clear something went wrong.

Reply:
Not an actual problem for this code but it's better to use in64 for PTS and DTS.  If mp4ff uses uint64 it  will cause problems when used with ffmpeg libs - they use negative values as special (for example NO_PTS is negative int64).